### PR TITLE
Enhance command controller

### PR DESCRIPTION
**Pull request by katrybacka**

### CODE REVIEW

**Comment by katrybacka** [comment]
recheck

File: pkg/controller/command/command_controller.go [hunk 1]
```diff
@@ -306,33 +285,99 @@ func (r *ReconcileCommand) Reconcile(request reconcile.Request) (reconcile.Resul
 		return reconcile.Result{}, err
 	}
 
-	configVolumeName := request.Name + "-" + instanceType + "-volume"
-	csrSignerCaVolumeName := request.Name + "-csr-signer-ca"
+	deployment := newDeploymentWithVolumes(command.Name, command.Namespace, instanceType, command.Spec.ServiceConfiguration.Containers)
+	expectedDeployment := deployment.DeepCopy()
+	performUpgrade := false
+	createOrUpdateResult, err := controllerutil.CreateOrUpdate(context.Background(), r.client, deployment, func() error {
+		if imageIsUpgraded(expectedDeployment, &deployment.Spec) || command.Status.UpgradeState == contrail.CommandStartingUpgradedDeployment {
```
[the remark this answers is not in the document]
Comment (kaweue) [reply pkg/controller/command/command_controller.go:292]: In my opinion
- prepareIntendedDeployment method should be called only in one place in code
- performUpgradeStep method shouldn't read deployment on its own (it looks like now we read it twice during the one reconciliation)
- performUpgradeStep shouldn't modify the deployment only state of Command CR

File: pkg/controller/command/command_upgrade.go [hunk 2]
```diff
@@ -16,40 +16,59 @@ import (
...
-		expectedReplicas := ptrToInt32(currentDeployment.Spec.Replicas, 1)
-		if currentDeployment.Status.ReadyReplicas == expectedReplicas {
-			commandCR.Status.UpgradeState = contrail.CommandNotUpgrading
+		expectedDeployment.Spec.DeepCopyInto(&d.Spec)
+		d.Spec.Replicas = command.Spec.CommonConfiguration.Replicas
+		if err := r.prepareIntendedDeployment(d, command); err != nil {
+			return nil, err
+		}
+		expectedReplicas := ptrToInt32(d.Spec.Replicas, 1)
+		if d.Status.ReadyReplicas == expectedReplicas {
```
Comment (kaweue) [inline pkg/controller/command/command_upgrade.go:61]: Is it possible that we will have more than one ReadyReplica in that state?

File: pkg/controller/command/command_upgrade.go [hunk 3]
```diff
@@ -76,6 +95,19 @@ func getContainerImage(containers []core.Container, name string) string {
 	return ""
 }
 
+func (r *ReconcileCommand) checkIfImageChangedAgain(command *contrail.Command) (bool, error) {
```
Comment (kaweue) [inline pkg/controller/command/command_upgrade.go:98]: Maybe information about to which version we are upgrading should be in the status of Command.

File: pkg/controller/command/command_upgrade.go [hunk 4]
```diff
@@ -76,6 +95,19 @@ func getContainerImage(containers []core.Container, name string) string {
...
+	}
+	if err != nil {
+		return false, err
+	}
+	return dataMigrationJob.Spec.Template.Spec.Containers[0].Image != getImage(command.Spec.ServiceConfiguration.Containers, "api"), nil
+}
+
 func (r *ReconcileCommand) checkDataMigrationJobStatus(command *contrail.Command) error {
 	dataMigrationJob := &batch.Job{}
 	jobName := types.NamespacedName{Namespace: command.Namespace, Name: command.Name + "-upgrade-job"}
```
Comment (kaweue) [inline pkg/controller/command/command_upgrade.go:113]: Should this "check" method create migrationJob `r.client.Create(context.Background(), dataMigrationJob)`

File: pkg/controller/command/command_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaweue) [reply pkg/controller/command/command_controller.go:292]: [Optional idea]
Moreover I would also go into completely deferent approach. More object oriented one.
Object CommandCR could have a method/methods that would be responsible for changing the inner state of object. Something like:
- reconcileUpgradeState(deploymentImage string) - that would internally change the state of CR
- upgradeJobComplated()
- getDeploymentSpec()
- getUpgradeState()

**Review by kaweue — CHANGES_REQUESTED** [review]

File: pkg/controller/command/command_upgrade.go (hunk 4 above)
Comment (katrybacka) [reply pkg/controller/command/command_upgrade.go:113]: no it should not, copy paste mistake

File: pkg/controller/command/command_upgrade.go (hunk 2 above)
Comment (katrybacka) [reply pkg/controller/command/command_upgrade.go:61]: yes, we are in that state until whole deployment is ready

File: pkg/controller/command/command_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (katrybacka) [reply pkg/controller/command/command_controller.go:292]: i'm not sure about this idea of not modifying deployment in performupgrade state just command cr, i have to think about that. i'll think about splitting logic of modifying cr and deployment

**Review by kaweue — APPROVED** [review]

**Review by kabart — APPROVED** [review]

**Pull request merged**